### ISSUE TITLE
[strict-boolean-expressions] Option to enable only in if, while, do-while and ternary

### ISSUE THREAD

**Issue by astoilkov** [typescript-eslint/typescript-eslint#903]
I am proposing an option to allow non-boolean expressions when outside if, while, do-while and ternary. I believe this is useful because it is way easier to read `foo || bar` than `foo === undefined ? bar : foo` and when there are more than two values things get complicated - `foo || bar || baz`.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": ["error"]
  }
}
```

```TS
const value = foo || bar
```

**Expected Result**

Currently, it produces an error.

**Actual Result**

To not produce an error if the new option is turned on.

**Additional Info**

I know you are aiming for TSLint feature parity first but after I turned this rule and fixed around 400 errors in our codebase I found out it is extremely useful in conditions and somewhat hard to fix(code becomes less readable) in other expressions.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.0.0` |
| `@typescript-eslint/parser`        | `2.0.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.2.2` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.9.0` |

**Comment by astoilkov** [comment]
After I though a little more about this. Maybe it is possible the option to only skip expressions part of assignments.

**Comment by phaux** [comment]
There's already `ignoreRhs` option which does that

**Comment by astoilkov** [comment]
@phaux Hmm, I have `ignoreRhs` turned on and the code below produces an error for the rule:
```ts
const foo = undefined
const bar = {}
const baz = foo || bar // Unexpected non-boolean in conditional.
```

From what I have seen `ignoreRhs` won't produce an error if the `foo` variable is a boolean but the `bar` variable remains an object:
```ts
const foo = undefined
const bar = {}
const baz = foo || bar // No error with `ignoreRhs: true`
```

**Comment by astoilkov** [comment]
@bradzacher Thanks for the reply. I like your argument with developers finding their way around rules. I am trying to find a solution to the problem which is elegant so thanks for your argument.

From looking at the TSLint options I want to ask if `allow-null-union` option will make this code not error?
```ts
function getPattern(line: string) {
  const unorderedListRegEx = /^\s*[*+-]\s+/u
  const orderedListRegEx = /^\s*(\d{1,9})[.)]\s+/u
  const taskListRegEx = /^(\s*[*+-]\s+)(\[[x X]\]\s+)/u
  const quoteRegEx = /^([ ]*>[ ]*\s+)/u
  const match =
      line.match(taskListRegEx) ||
      line.match(unorderedListRegEx) ||
      line.match(orderedListRegEx) ||
      line.match(quoteRegEx)
}
```

**Comment by astoilkov** [comment]
@bradzacher Thank you so much for the response. This gave so much clarity to my questions. I didn't answer because I didn't receive an email notification for your answer. I accidentally returned to the thread to remember what happened as the new `allowNullable` option was introduced. You are awesome!
